tools/packages/wesl/src/VirtualFilesystem.ts

```diff
@@ -0,0 +1,89 @@
+import { assertThat } from "./Assertions";
+import { isIdent } from "./parse/WeslStream";
+
+/**
+ * A relative, normalized, Linux-style path ending with a file.
+ * All file and folder names must be valid WGSL identifiers.
+ * `.wgsl` and `.wesl` are the supported file extensions.
+ */
+export type WeslPath = string & { __weslPath: never };
+
+/** A async virtual filesystem, can be backed by an in-memory map, or by a real filesystem, or by HTTP requests */
+export interface VirtualFilesystem {
+  /**
+   * The WESL module loading uses this to load files.
+   */
+  readFile(path: WeslPath): Promise<string | null>;
+}
+
+/** Creates a static filesystem from relative, normalized, Linux-style paths.
+ * All file and folder names must be valid WGSL identifiers.
+ * `.wgsl` and `.wesl` are the supported file extensions.
+ */
+export function staticFilesystem(
+  files: Record<string, string>,
+): VirtualFilesystem {
+  // Avoid accidentally having a file called `toString`
+  const filesystemMap = new Map(
+    Object.entries(files).map(([path, contents]) => [
+      makeWeslPath(path),
+      contents,
+    ]),
+  );
+  function readFile(path: WeslPath): Promise<string | null> {
+    const file = filesystemMap.get(path) ?? null;
+    return Promise.resolve(file);
+  }
+  return {
+    readFile,
+  };
+}
+
+const fileNameRegex = /^(?<name>[^.]+)\.(?<extension>wgsl|wesl)$/;
+
+export function makeWeslPath(path: string): WeslPath {
```

Comment:
Doing lots of checks here is a good idea, since we want to tell the user about screwups as early as possible.
And then internally we get to assume that all paths have some very simple properties, instead of dealing with the insanity of real filesystems.